jibby.go

```diff
@@ -0,0 +1,529 @@
+// readNameSeparator expects the '}' character after optional white space and
+// errors if it is not found.  It handles other errors like readCharAfterWS.
+func (d *Decoder) readObjectTerminator() error {
+	err := d.readCharAfterWS('}')
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// readNameSeparator expects the '"' character after optional white space and
+// errors if it is not found.  It handles other errors like readCharAfterWS.
+func (d *Decoder) readQuoteStart() error {
+	err := d.readCharAfterWS('"')
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// readSpecificKey expects and consumes a specific series of bytes representing
+// an object key, and errors if it is not found.  The starting quote character
+// must have already been consumed from the input stream.  The closing quote and
+// subsequent name separator will also be consumed from the input stream.  It
+// handles other errors like readCharAfterWS.
+func (d *Decoder) readSpecificKey(expected []byte) error {
+	charsNeeded := len(expected) + 1
+	key, err := d.peekBoundedQuote(charsNeeded, charsNeeded)
+	if err != nil {
+		return fmt.Errorf("expected %q, but failed to find key: %v", string(expected), err)
+	}
+	if !bytes.Equal(key, expected) {
+		if len(key) > 0 {
+			return d.parseError(key[0], fmt.Sprintf("expected %q", string(expected)))
+		}
+		return d.parseError('"', fmt.Sprintf("expected %q", string(expected)))
+	}
+	_, _ = d.json.Discard(len(key) + 1)
+	err = d.readNameSeparator()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// peekNumber peeks into the input stream and returns a slice that might be
+// parsable as a number, a boolean hint whether it should be treated as a
+// floating point number, and any error that is found.  The input stream is
+// not consumed.
+//
+// The slice returned will contain characters that could appear in a JSON
+// floating point number (excluding "NaN" and "Inf", which are not legal JSON)
+// without regard to legal arrangment.  For example, `123-eee` is a possible
+// return value.  This function only only parses out a string to be passed to a
+// numeric conversion function.
+func (d *Decoder) peekNumber() ([]byte, bool, error) {
+	var isFloat bool
+	var terminated bool
+
+	buf, err := d.json.Peek(doublePeekWidth)
+	if err != nil {
+		// here, io.EOF is OK, since we're peeking and may hit end of
+		// object within the peek width.
+		if err != io.EOF {
+			return nil, false, newReadError(err)
+		}
+	}
+
+	// Find where the number appears to ends and if it's a float.  A
+	// number ends at white space, a separator, or a terminator.
+	var i int
+LOOP:
+	for i = 0; i < len(buf); i++ {
+		switch buf[i] {
+		case 'e', 'E':
+			isFloat = true
+		case '.':
+			isFloat = true
+			if i < len(buf)-1 && (buf[i+1] < '0' || buf[i+1] > '9') {
+				return nil, false, d.parseError(buf[i], "decimal must be followed by digit")
+			}
+		case ' ', '\t', '\n', '\r', ',', ']', '}':
+			terminated = true
+			break LOOP
+		case '_':
```

Comment:
Why is `'_'` special? Isn't _any_ other character invalid? I'm not sure why this method would return an error for `"1_2"` and not `"1a2"`.